mdtk/midi.py

```diff
@@ -0,0 +1,114 @@
+"""Code to interact with MIDI files, including parsing and converting them
+to csvs."""
+
+import os
+import glob
+import pandas as pd
+
+import pretty_midi
+
+
+
+
+def midi_dir_to_csv(midi_dir_path, csv_dir_path):
+    """
+    Convert an entire directory of MIDI files into csvs in another directory.
+    This searches the given MIDI path for any files with the extension 'mid'.
+    It will create any necessary directories for the csv files, and will
+    create one csv per MIDI file, where the 'mid' extension is replaced with
+    'csv'.
+    
+    Parameters
+    ----------
+    midi_dir_path : string
+        The path of a directory which contains any number of MIDI files with
+        extension 'mid'. The directory is not searched recursively, and any
+        files with a different extension are ignored.
+        
+    csv_dir_path : string
+        The path of the directory to write out each csv to. If it does not
+        exist, it will be created.
+    """
+    for midi_path in glob.glob(midi_dir_path + os.path.sep + '*.mid'):
+        csv_path = (csv_dir_path + os.path.sep +
+                    os.path.basename(midi_path[:-3] + 'csv'))
+        midi_to_csv(midi_path, csv_path)
+
+
+
+def midi_to_csv(midi_path, csv_path):
+    """
+    Convert a MIDI file into a csv file.
+    
+    Parameters
+    ----------
+    midi_path : string
+        The filename of the MIDI file to parse.
+        
+    csv_path : string
+        The filename of the csv to write out to. Any nested directories will be
+        created by df_to_csv(df, csv_path).
+    """
+    df_to_csv(midi_to_df(midi_path), csv_path)
+    
+    
+    
+    
+def midi_to_df(midi_path):
+    """
+    Get the data from a MIDI file and load it into a pandas DataFrame.
+    
+    Parameters
+    ----------
+    midi_path : string
+        The filename of the MIDI file to parse.
+        
+    Returns
+    -------
+    df : DataFrame
+        A pandas DataFrame containing the notes parsed from the given MIDI
+        file. There will be 4 columns:
+            onset: Onset time of the note, in milliseconds.
+            track: The track number of the instrument the note is from.
+            pitch: The MIDI pitch number for the note.
+            dur: The duration of the note (offset - onset), in milliseconds.
+        Sorting will be first by onset, then track, then pitch, then duration.
+    """
+    midi = pretty_midi.PrettyMIDI(midi_path)
+    
+    notes = []
+    for index, instrument in enumerate(midi.instruments):
+        for note in instrument.notes:
+            notes.append({'onset': note.start * 1000,
+                          'track': index,
```

Comment:
There's not really a way to ensure this as far as I know, since different datasets probably have different ways of treating each instrument.

There seems to be a [MIDI standard](https://en.wikipedia.org/wiki/General_MIDI#Program_change_events), but that refers to actual MIDI events, independent of track number (and this is not guaranteed to be consistent even within a single piece). There may also be, eg. 2 separate guitar parts but on different tracks. The closest thing to a guarantee I've seen is that track 10 tends to be percussion.

So in short: no, it doesn't ensure that, and it would likely be non-trivial (but possible) to do so.

Reply:
Ok, so it seems pretty_midi ([here](https://github.com/craffel/pretty-midi/blob/78e33ab1c24e1156529320ea127dd51d57f7ab99/pretty_midi/pretty_midi.py#L226)) separates by "instrument,track" pair.

There are some MIDI files which separate into tracks by voice (for example fugues). So those would come as separate "instruments" for us (which is probably what we want). I'd go for leaving this alone for now.

Reply:
Sure - I agree: we'll come back to this if required